src/track/beatutils.cpp

```diff
@@ -13,49 +13,181 @@
+        tempoFrequency, tempoList);
+    QVector<double> fixedBeats;
+    QMap<int, double> fixedTempoAndBeats;
+    // here we only care about where the change happened
+    auto tempoChanges = stableTemposAndPositions.keys();    
+    for (int lastTempoChage = 0; 
+                lastTempoChage < tempoChanges.count() -1;
+                lastTempoChage++) {
+
+        int beatStart = tempoChanges[lastTempoChage];
+        int beatEnd = tempoChanges[lastTempoChage + 1];
+        int beatDiff = beatEnd - beatStart;
+        // here we detect if the segment has a constant tempo or not
+        double bpmDiff = 0;
+        if (beatDiff > kBeatsToCountTempo) {
+            int middle = beatDiff / 2;            
+            QVector<double> beatsAtLeft(
+                    rawBeats.begin() + beatStart, rawBeats.begin() + beatStart + middle);
+            QVector<double> beatsAtRight(
+                    rawBeats.begin() + beatStart + middle, rawBeats.begin() + beatEnd);
+            // the median is a better guess of similarity than our more
+            // sophisticated method calculatebpm that might consider outliers
+            double beginBpm = BeatUtils::medianBpm(beatsAtLeft, SampleRate, minBpm, maxBpm);
+            double endBpm = BeatUtils::medianBpm(beatsAtRight, SampleRate, minBpm, maxBpm);
+            bpmDiff = fabs(beginBpm - endBpm);
+        }
+        // here we handle ramping or unsteady values by making a beatgrid for each bar
+        if (bpmDiff >= BPM_ERROR) {
+            // TODO(Cristiano) change hardcoded 4s for detected bpb
+            while (beatStart < beatEnd - 4) {
+                QVector<double> splittedAtMeasure(
+                        rawBeats.begin() + beatStart, rawBeats.begin() + beatStart + 4);
+                // for such a small vector the median is also more realible
+                double measureBpm = medianBpm(splittedAtMeasure, SampleRate, minBpm, maxBpm);
+                fixedTempoAndBeats[beatStart] = measureBpm;
+                fixedBeats << BeatUtils::calculateFixedTempoBeatMap(
+                        splittedAtMeasure, SampleRate, measureBpm, false);
+                beatStart += 4;
+            }
+        } else {
+            // here we have only one bpm for whole segment
+            QVector<double> splittedAtTempoChange(
+            rawBeats.begin() + beatStart, rawBeats.begin() + beatEnd);
+            double bpm = calculateBpm(splittedAtTempoChange, SampleRate, minBpm, maxBpm);
+            fixedBeats << BeatUtils::calculateFixedTempoBeatMap(
+                    splittedAtTempoChange, SampleRate, bpm);
+            fixedTempoAndBeats[beatStart] = bpm;
+        }
+    }
+    return std::make_tuple(fixedBeats, fixedTempoAndBeats);
+}
 
-        local_bpm = floor(local_bpm * kHistogramDecimalScale + 0.5) / kHistogramDecimalScale;
-        frequency[local_bpm] += 1;
+void BeatUtils::printBeatStatistics(const QVector<double>& beats, int SampleRate) {
```

Comment:
Why not aggregate the statistics into a separate structure and return it? This structure could then be printed on demand, i.e. separate the functional calculation from the I/O side effect.